lib/file-source/src/fingerprinter.rs

```diff
@@ -69,6 +70,74 @@ impl From<u64> for FileFingerprint {
     }
 }
 
+#[derive(Debug, Copy, Clone)]
+enum SupportedCompressionAlgorithms {
+    GZIP,
+}
+
+impl SupportedCompressionAlgorithms {
+    fn values() -> Vec<SupportedCompressionAlgorithms> {
+        vec![SupportedCompressionAlgorithms::GZIP]
+    }
+
+    fn magic_header_bytes(&self) -> &'static [u8] {
+        match self {
+            SupportedCompressionAlgorithms::GZIP => &[0x1f, 0x8b],
+        }
+    }
+}
+
+trait UncompressedReader {
+    fn check(fp: &mut File) -> Option<SupportedCompressionAlgorithms>;
+    fn reader<'a>(fp: &'a mut File) -> Box<dyn BufRead + 'a>;
+}
+
+struct UncompressedReaderImpl;
+impl UncompressedReader for UncompressedReaderImpl {
+    fn check(fp: &mut File) -> Option<SupportedCompressionAlgorithms> {
+        SupportedCompressionAlgorithms::values()
+            .iter()
+            .find_map(|compression_algorithm| {
+                // magic headers for algorithms can be of different lengths, and using a buffer too long could exceed the length of the file
+                // so instantiate and check the various sizes independently
+                let magic_header_bytes = compression_algorithm.magic_header_bytes();
+                let mut magic = vec![0u8; magic_header_bytes.len()];
+
+                if fp.read_exact(&mut magic).is_ok()
```

Comment:
If this fails, I think it could be due to one of two reasons:

- Not enough bytes
- Read interrupted

If the read is interrupted, I think we likely want to return an error since we weren't able to detect whether the file is compressed.

If there aren't enough bytes, we probably want to try again later since even uncompressed files, I think, should be a minimum of two bytes (a character and a newline) 🤔 The risk is that the file is actually compressed, there just aren't enough bytes to determine that yet.

Reply:
Errors from `read_exact` are documented here: https://doc.rust-lang.org/std/io/trait.Read.html#method.read_exact

Reply:
Thanks @jszwedko , I appreciate the review. I think I follow most of your points

I guess it would make sense most to:
* Check each magic bytes header in the file. If find one that matches an algorithm, then short-circuit and return the corresponding enum
* If **all** header checks fail, it means that this file is either:
    *  Uncompressed, in which case we should return `None`, so we can fingerprint as-is
    *  Using an unsupported compression algorithm, in which case we should return `None`, so we can fingerprint as-is
    * Using a supported compression algorithm, but the read was interrupted--return an error if read_exact fails for `[ErrorKind::Interrupted]`.
    * Using a supported compression algorithm, but there aren't enough bytes to determine that yet 
        * So in this scenario, I guess you're suggesting this is some kind of compressed stream that doesn't have a header written to it yet, and we should throw an error, as this would manifest as `[ErrorKind::UnexpectedEof]`. 
        * There is one tiny edge case here that will cause a potentially valid assessment to instead fail, which is if the compression algorithm we're checking happens to have a bigger header than the intended compression algorithm. e.g.) zstd's magic bytes are 4 bytes, assume the file is 3 bytes gzipped (gzip headers are 2 bytes, 1 byte of compressed content). If we throw an error at zstd before checking gzip, then we'd needlessly try again.
        *  I guess this can be prevented by deterministically checking from smallest compression algorithms to the largest.

Lastly:
>we probably want to try again later

**I presume this is covered up the stack if fingerprinting fails?** I can investigate myself, but if you happen to have the answer handy, it'd be helpful